new section for the migration docs (#17)
onmax/nimiq-developer-center: migration/node-operators.md

# PoS Migration Guide for Node Operators

Our migration guides are designed for users who want to actively participate in the transition process. This specific guide is for those who want to run the migration process without becoming validators (if you want to become one of the first validators from the start of the PoS chain, please refer to this [guide](migration-validators)). It is suitable for:

- Node Operators
- Exchanges
- Regular Users

This guide provides instructions to run the Activation Tool, the final step of the migration to the Nimiq PoS. There is no requirement to run the migration as a node operator. You can simply start your PoS client once the migration process ends and build your own node.

> It's important to note that, as a node operator, your role is primarily as an observer during the activation process. The migration will only proceed if at least 80% of the allocated stake is ready to migrate. Since you have not pre-registered as a validator, running the activation tool does not contribute to meet the threshold requirement for the transition. Nonetheless, by running the migration tool, you will replicate the state migration and genesis block creation process.

Reviewer: The advantage of running the activation tool as a node operator is to have 0% downtime and being able to observe the whole transition process but is not mandatory.

> It's important to note that, as a node operator, your role is primarily as an observer during the activation process. The migration will only proceed if at least 80% of the allocated stake is ready to migrate. Since you have not pre-registered as a validator, running the activation tool does not contribute to meeting the threshold requirement for the transition. Nonetheless, by running the migration tool, you will participate in the state migration and genesis block creation process.

Reviewer 2: ```suggestion
It's important to note that, as a node operator, your role is primarily as an observer during the activation process. The migration will only proceed if at least 80% of the allocated stake is ready to migrate. Since you have not pre-registered as a validator, running the activation tool does not contribute to meet the threshold requirement for the transition. Nonetheless, by running the migration tool, you will replicate the state migration and genesis block creation process.
```

::: details Users without an Existing PoW Client

- Clone the [Nimiq CoreJS repository](https://github.com/nimiq/core-js?tab=readme-ov-file#quickstart) and follow the instructions
- Refer to [this sample guide](https://github.com/nimiq/core-js/blob/master/clients/nodejs/sample.conf) to enable the PoW RPC server
- Clone the [PoS blockchain repository](https://github.com/nimiq/core-rs-albatross?tab=readme-ov-file#installation) and follow the instructions to compile the code

**Step 1: Set Up your PoS Configuration File**

This step involves configuring your PoS client settings to prepare for the transition. You'll set parameters such as the network, sync mode, and RPC server.

Once you have your client compiled in the **PoS chain**, you can:

- Open the file `$HOME/.nimiq/client.toml` (see [example](https://github.com/nimiq/core-rs-albatross/blob/albatross/lib/src/config/config_file/client.example.toml))
- Set the `network` to `main`
- Choose your `sync_mode` setting as `full` or `history`

> - Enable the PoS RPC server by uncommenting the section in the `[rpc-server]` configuration

Reviewer 2: This is actually not required and should be optional.

**Step 2: Check Client Sync Status**

**In the PoW chain**, ensure you are fully synced and in consensus. Start the PoW client with an RPC server (this might take a while). Run the following command:

```bash
node clients/nodejs/index.js --dumb --network=main --rpc=8648
```

**Step 3: Run the Activation Tool**

> The Activation Tool establishes a connection with the PoW chain via RPC, extracting data from your PoW client with the PoS client configuration.

Reviewer 2: ```suggestion
The Activation Tool establishes a connection with the PoW client via RPC, extracting data from your PoW client with the PoS client configuration.
```

> Ensure your PoW client is fully synced before running the Activation Tool on the PoS chain side. Before executing the Activation Tool, compile the **PoS chain** code by running `cargo build --release`.

Reviewer 2: ```suggestion
Ensure your PoW client is fully synced before running the Activation Tool on the PoS chain side. Before executing the Activation Tool, compile the **PoS chain** client by running `cargo build --release`.
```

> Once you are in consensus in the PoW chain, proceed to execute the Activation Tool by running the following command **in the PoS chain**, including the path to the configuration file containing your validator data and specifying the RPC server to be used:

Reviewer 2: ```suggestion
Once you are in consensus in the PoW chain, proceed to execute the Activation Tool by running the following command **in the PoS client repository**, including the path to the configuration file containing your validator data and specifying the RPC server to be used:
```

```bash

> cargo run --release --bin nimiq-pow-migration --url "http://127.0.0.1:8648" --config client.toml

Reviewer 2: This is assuming that the `PoW` client runs in the same machine as the `PoS` client. This might not be true for some people. Could we clarify this? Also I think some cases might use `https` instead.

```

:::

::: details Users with an Existing PoW Client

- Make sure you have your PoW RPC server enabled to allow your PoS client to connect to it
- Clone the [PoS blockchain repository](https://github.com/nimiq/core-rs-albatross?tab=readme-ov-file#installation) and follow the instructions to compile the code

> **Step 1: Set Up your PoS Configuration File**

Reviewer 2: This was described above... Copy/paste error?

Author: this guide has 2 guides. one for nodes that already running and the other for users that want to run a node for the migration

This step involves configuring your PoS client settings to prepare for the transition. You'll set parameters such as the network, sync mode, and RPC server.

Once you have your client compiled, you can:

- Open the file `$HOME/.nimiq/client.toml` (see [example](https://github.com/nimiq/core-rs-albatross/blob/albatross/lib/src/config/config_file/client.example.toml))
- Set the `network` to `main`
- Choose your `sync_mode` setting as `full` or `history`

> - Enable the PoS RPC server by uncommenting the section in the `[rpc-server]` configuration

Reviewer 2: This step is optional and is not mandatory.

**Step 2: Run the Activation Tool**

The Activation Tool establishes a connection with the PoW chain via RPC, extracting data from your PoW client with the PoS client configuration.

Ensure your PoW client is fully synced before running the Activation Tool on the PoS chain side. Before executing the Activation Tool, compile the **PoS chain** code by running `cargo build --release`.

> Once you are in consensus in the PoW chain, proceed to execute the Activation Tool by running the following command **in the PoS chain**, including the path to the configuration file containing your validator data and specifying the RPC server to be used:

Reviewer 2: ```suggestion
Once you are in consensus in the PoW chain, proceed to execute the Activation Tool by running the following command **in the PoS client repository**, including the path to the configuration file containing your validator data and specifying the RPC server to be used:
```

```bash

> cargo run --release --bin nimiq-pow-migration --url "http://127.0.0.1:8648" --config client.toml

Reviewer 2: Same, this is assuming that the `PoS` client and server run in the same machine. And I guess some cases might be `https` instead.

```

:::

> The Activation Tool connects with the PoW chain via RPC to extract necessary data for the migration process. After launching the tool, it proceeds as follows:

Reviewer 2: ```suggestion
The Activation Tool connects with the PoW client via RPC to extract necessary data for the migration process. After launching the tool, it proceeds as follows:
```

> - Selects the block for the transition to the PoS chain

Reviewer 2: ```suggestion
- Selects the PoW block for the transition to the PoS chain
```

- Starts the migration of the state from the PoW chain to the PoS chain
- Builds the genesis block for the PoS chain
- Monitors the readiness of validators, waiting for at least 80% of them to signal readiness

> Once this readiness threshold is met, the PoS client is launched with the new genesis block, marking the beginning of the PoS chain. Mind that as a non-validator, your role during this process is primarily observant, as your participation does not contribute to reaching the 80% readiness threshold.

Reviewer 2: ```suggestion
Once this readiness threshold is met, the PoS client is launched with the new genesis block, marking the beginning of the PoS chain. Mind that as a non-validator, your role during this process is primarily observant, as your participation does not contribute to reach the 80% readiness threshold.
```

- Open the file `$HOME/.nimiq/client.toml` (see [example](https://github.com/nimiq/core-rs-albatross/blob/albatross/lib/src/config/config_file/client.example.toml))
- Set the `network` to `main`
- Choose your `sync_mode` setting as `full` or `history`

> - Optionally, enable the PoS RPC server by uncommenting the section in the [rpc-server] configuration if you need it.

Reviewer 2: For consistency with regards the other bullet points:
```suggestion
- Optionally, enable the PoS RPC server by uncommenting the section in the [rpc-server] configuration if you need it
```
